xhydro/optimal_interpolation/optimal_interpolation_fun.py

```diff
@@ -0,0 +1,571 @@
+
+    # return the flow quantiles as desired.
+    return flow_quantiles
+
+
+def execute_interpolation(
+    qobs: xr.Dataset,
+    qsim: xr.Dataset,
+    station_correspondence: xr.Dataset,
+    crossvalidation_stations: list,
+    ratio_var_bg,
+    percentiles,
+    iterations,
+    write_file,
+    parallelize: bool = False,
+    max_cores: int = 1,
+):
+    """Execute the main code, including setting constants to files, times, etc.
+
+    Parameters
+    ----------
+    qobs : xr.Dataset
+        Streamflow and catchment properties dataset for observed data.
+    qsim : xr.Dataset
+        Streamflow and catchment properties dataset for simulated data.
+    station_correspondence : xr.Dataset
+        Matching between the tag in the HYDROTEL simulated files and the observed station number for the obs dataset.
+    crossvalidation_stations : list
+        Observed hydrometric dataset stations to be used in the cross-validation step.
+    ratio_var_bg : float
+        Ratio for background variance.
+    percentiles : list
+        Desired percentiles for flow quantiles.
+    iterations : int
+        The number of iterations for the interpolation.
+    write_file : str
+        Name of the NetCDF file to be created.
+    parallelize : bool
+        Flag indicating whether to parallelize the interpolation.
+    max_cores : int
+        Maximum number of cores to use for parallel processing.
+
+    Returns
+    -------
+    list
+        A list containing the flow quantiles for each desired percentile.
+    """
+    filtered_dataset = retrieve_data(
+        qobs=qobs,
+        qsim=qsim,
+        station_correspondence=station_correspondence,
+        crossvalidation_stations=crossvalidation_stations,
+    )
+
+    # Project centroid lat/longs in a plane to get distances in km
+    x, y = ma.latlon_to_xy(
+        filtered_dataset["centroid_lat"].values,
+        filtered_dataset["centroid_lon"].values,
+        np.array([45] * len(filtered_dataset["centroid_lat"].values)),
+        np.array([-70] * len(filtered_dataset["centroid_lat"].values)),
+    )
+
+    x_points, y_points = standardize_points_with_roots(
+        x,
+        y,
+        len(filtered_dataset["station_id"].values),
+        filtered_dataset["drainage_area"].values,
+    )
+
+    # create the weighting function parameters
+    ecf_fun, par_opt = ecf_cc.correction(
+        filtered_dataset["qobs"].values,
+        filtered_dataset["qsim"].values,
+        x_points,
+        y_points,
+        iteration_count=iterations,
+    )
+
+    flow_quantiles = parallelize_operation(
+        filtered_dataset=filtered_dataset,
+        ecf_fun=ecf_fun,
+        par_opt=par_opt,
+        x_points=x_points,
+        y_points=y_points,
+        ratio_var_bg=ratio_var_bg,
+        percentiles=percentiles,
+        iterations=iterations,
+        parallelize=parallelize,
+        max_cores=max_cores,
+    )
+
+    # Write results to netcdf file
+    util.write_netcdf_flow_percentiles(
```

Comment:
```suggestion

    # TODO: Reassess if we really need to write the results here.
    # Write results to netcdf file
    util.write_netcdf_flow_percentiles(
```
Je ne sais pas si l'action de sauvegarder devrait être faite à-même `xhydro`. Le plus possible, je pense qu'on vise de retourner des datasets et laisser la sauvegarde à l'utilisateur. Cela dit, pour l'instant on peut garder ça comme ça.

Reply:
J'ai changé tel que discuté, on retourne maintenant un dataset qu'on peut traiter comme on veut.